### ISSUE TITLE
Enumerable::EmptyError on Hash#reduce

### ISSUE THREAD

**Issue by stugol** [crystal-lang/crystal#2155]
This code throws an unnecessary `Enumerable::EmptyError`:

```
a = Array(Hash(String, String)).new
a.reduce { |acc, v| 
  {} of String => String
}
```

**Comment by asterite** [comment]
Please explain why, what it should return instead and what's the type of `v`. If it should not raise, provide a pull request that fixes it.

**Comment by stugol** [comment]
Well, because reducing an empty array should not be an error.

Whether the block should actually be invoked in the absence of any values, is a question for the development team - that is, you - to decide. If so, `v` should probably be of type `T | Nil`. If not, then the result of `reduce` should probably be `U | Nil`, where U is whatever the block returns. I would lean towards the latter.

Perhaps, to preserve backwards compatibility, a new `reduce?` function should be added; while the original remains unchanged, except maybe to improve the error message. `Some part of your program threw an EmptyError - have fun finding it!` is not helpful, as it is only meaningful to people who know [a] where the exception originates, and [b] how `reduce` actually works internally.

I find your response somewhat dismissive. Did you actually _intend_ it to sound like "if you want it to work properly, fix it yourself"? Because last I checked, I wasn't a Crystal dev. If you'd like me to contribute, might I suggest _asking_ me to, rather than ordering me to?

**Comment by stugol** [comment]
Upon consideration, it seems to me that the possibility of an empty sequence is a risk that every call to `reduce()` takes. By asking for a sequence reduction **without providing an initial value**, you are _de facto_ accepting that the answer may be "nothing".

Therefore, I propose that the result of calling `T#reduce()` should always be `U | Nil`. If the programmer wants a guarantee of a non-nil result, he should use the `T#reduce(U)` overload.

I would be happy to attempt a pull request if this meets with your approval....?

**Comment by bcardiff** [comment]
So, the discussing is about how reduce should play when non initial value is given. In other scenarios there is no disagreement.

a) Raise non EmptyError
b) Return Nil

Raising is similar to `foldr1` [in Haskell](http://hackage.haskell.org/package/base-4.8.2.0/docs/Prelude.html#v:foldr1). There is no other foldr like function without initial value.

So I think is ok for the default `reduce` to raise exception.
I also think that the most common way to call `reduce` is with an initial value.

But I don't see any harm of adding a `reduce?` alternative that will conform with what @stugol suggest, i.e. returning nil on empty sequence.

**Comment by stugol** [comment]
Consider the exception produced:
```
Empty enumerable (Enumerable::EmptyError)
[134603337] *CallStack::unwind:Array(Pointer(Void)) +121
[134603199] *CallStack#initialize<CallStack>:Array(Pointer(Void)) +31
[134603141] *CallStack::new:CallStack +69
[134703944] *Exception +56
[134703852] *Enumerable::EmptyError#initialize<Enumerable::EmptyError, String>:CallStack +60
[134703771] *Enumerable::EmptyError::new<String>:Enumerable::EmptyError +139
[134703616] *Enumerable::EmptyError::new:Enumerable::EmptyError +48
[134588510] ???
[134596557] main +77
[3074761879] __libc_start_main +247
[134585089] ???
```
The stack trace doesn't even mention the `reduce` call; neither does it have line numbers. Tracking down a Crystal exception is a long and arduous process. Doesn't it make sense, therefore, to avoid throwing exceptions when there is another reasonable alternative?

On that subject, I propose that every instantiation of an exception be rewritten by the compiler to pass the file and line number to the constructor. At least then, we might have a hope in hell of tracking it down. The way things stand, it might as well just output "something went wrong somewhere".

**Comment by bcardiff** [comment]
I agree the exception trace need some improvements. Knowing where it was raised is a useful almost mandatory thing, but when a method with a block is called a specialized inlined version of the method is created in order to avoid creating a closure for the block and successive jumps. That's the source of conflict here with stack information. But all in all is a different issue, that is about improving exception traces for methods with blocks, which I bet we all agree we want better stuff there.

But in order to try to reach an agreement regarding the reduce semantic, do you agree with my previous statement?

As for exception trace, we can start another discussion on that to deal with internals as see who and when can invest effort there.

**Comment by stugol** [comment]
@bcardiff Personally, I think the parameterless overload should just return `T | Nil`. It makes more sense to me. Enumerations are _allowed_ to be empty, so the question "what is the sum of all these integers?" (given an empty list) should be either `0` or `nil`, not a meltdown.

**Comment by bcardiff** [comment]
It is the same issue as with `#first`/`#first?` [in Enumerable](http://crystal-lang.org/api/Enumerable.html#first-instance-method), as I suggested, I see no harm to follow and accept the same pattern with a `#reduce?(...) : T | Nil` overload.

If the default behavior injects `T | Nil` a lot of `#not_nil!` and casts will appear in the code and when the assumption of non empty is invalid a stack trace will be thrown. Me and others like the default behavior being to avoid Nil types as much as possible. And we should follow the same criteria (whatever it is) everywhere. 

With a `#not_nil!` or cast a stacktrace will be better since there is no block involved, but ~~if~~ when the stacktrace is improved for block versions there will be no difference.

If you program with `...?` versions of Enumerable methods I bet you will reach a lot of annoying `#not_nil!` and casts that will pollute the code. The unions and non nillable types is the best solution we've reached so far balancing expressiveness and safety.
